build_runner/lib/src/build/build_series.dart

```diff
@@ -243,5 +276,17 @@ class BuildSeries {
     return result;
   }
 
-  Future<void> beforeExit() => _resourceManager.beforeExit();
+  /// Ends the build series.
+  ///
+  /// First, any currently-running build is waited for.
+  ///
+  /// Then: resource cleanup is done, [buildResults] is closed and no further
+  /// builds are allowed.
+  Future<void> close() async {
+    if (_closedCompleter.isCompleted) return;
+    _closedCompleter.complete();
```

Comment:
Can't we get into a weird situation here where it is currently closing down but other parts think it is already done?

Reply:
Good point, if something waited for `closed` then called `exit(0)` that would interrupt cleanup.

Currently nothing does that, the one place it's used is in `watcher.dart` to decide to stop scheduling new builds.

So, renamed `closed` -> `closing`.

`watcher.dart` uses `onDone` on the build results stream to decide when to exit, which seems good. Moved closing the stream last in this method and added comments.